fundamentals/fundamentals-bio/src/main/scala/izumi/functional/bio/impl/BioEither.scala

```diff
@@ -0,0 +1,31 @@
+package izumi.functional.bio.impl
+
+import izumi.functional.bio.Error2
+
+import scala.util.Try
+
+object BioEither extends BioEither
+
+class BioEither extends Error2[Either] {
+  override def pure[A](a: A): Either[Nothing, A] = Right(a)
+  override def map[R, E, A, B](r: Either[E, A])(f: A => B): Either[E, B] = r.map(f)
+
+  /** execute two operations in order, map their results */
+  override def map2[R, E, A, B, C](firstOp: Either[E, A], secondOp: => Either[E, B])(f: (A, B) => C): Either[E, C] = {
+    firstOp.flatMap(a => secondOp.map(b => f(a, b)))
+  }
+  override def flatMap[R, E, A, B](r: Either[E, A])(f: A => Either[E, B]): Either[E, B] = r.flatMap(f)
+
+  override def catchAll[R, E, A, E2](r: Either[E, A])(f: E => Either[E2, A]): Either[E2, A] = r.left.flatMap(f)
+  override def catchSome[R, E, A, E1 >: E](r: Either[E, A])(f: PartialFunction[E, Either[E1, A]]): Either[E1, A] = r.left.flatMap(f)
+  override def fail[E](v: => E): Either[E, Nothing] = Left(v)
```

Comment:
catchSome - seems weird, but the rest should be OK

Reply:
@VladPodilnyk `catchSome` is incorrect, you should use `.applyOrElse` instead of passing f itself (it will throw exception when used as a function, instead it should do nothing if the error doesn't match the partial function)

Reply:
Note: cats law tests will probably show the error in `catchSome` (we use it to implement `cats.ApplicativeError#recoverWith`)